### PR TITLE
Switch benchmarking mode in `pytest-benchmark`

### PR DESCRIPTION
**Pull request by Priya2698**
The current benchmarks use the default mode in `pytest-benchmark` which determines the number of rounds and iterations the target function is run based on the minimum and maximum time allowed per benchmark. In this mode, smaller benchmarks can take a long time to run if the maximum time is not set properly.

This PR switches the benchmarks to use `benchmark.pedantic` which allows us to control the number of rounds each benchmark is run for. This change also supports running a setup function before each round that allows us to clear L2 cache everytime before the target function is executed instead of only before running the benchmark.
This is important for smaller input sizes where L2 cache impacts the measured timings.

For example: The following timings show that in the previous mode, the measured timings are shorter than the actual timings expected.

default_benchmark

With the pedantic mode, we get accurate results.

pedantic_benchmark

The exposed interface for running benchmarks remains same as detailed in PR #982 . The number of rounds/warmup rounds for the target function can be modified by passing the arguments `rounds`/ `warmup_rounds `to `run_benchmark`.

### CODE REVIEW

**Comment by Priya2698** [comment]
Should we have any warmup iterations? I did not notice any difference in timing measurements

**Comment by naoyam** [comment]
> Should we have any warmup iterations? I did not notice any difference in timing measurements with or without warmup.

I don't know if we need, but maybe in some cases? If not too complicated, it would be good to have an optional feature, i.e., should be turned off by default but could be enabled if desired.

**Review by naoyam — COMMENTED** [review]
LGTM. @jacobhinkle please review.

**Comment by Priya2698** [comment]
> > Should we have any warmup iterations? I did not notice any difference in timing measurements with or without warmup.
> 
> I don't know if we need, but maybe in some cases? If not too complicated, it would be good to have an optional feature, i.e., should be turned off by default but could be enabled if desired.


It is present as an option, the default value is zero. So we can turn them on for specific benchmarks when needed.

**Review by jacobhinkle — APPROVED** [review]
LGTM. Just one thing to consider re warmup, but I don't think we should need to warmup unless we start additionally measuring host time.

File: python_benchmarks/core.py [hunk 1]
```diff
@@ -193,7 +193,11 @@ def set_metrics(
 
 
 def run_benchmark(
-    benchmark, benchmark_fn: Callable, inputs: Union[torch.Tensor, List]
+    benchmark,
+    benchmark_fn: Callable,
+    inputs: Union[torch.Tensor, List],
+    rounds: int = 10,
+    warmup_rounds: int = 0,
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply python_benchmarks/core.py]: I didn't notice any difference but am setting warmup=1 to be safe.

**Comment by Priya2698** [comment]
We can potentially add command line options to set warmup rounds and benchmarking rounds. I'll add that in the future, will open an issue for now.

**Pull request merged**